### PR TITLE
Clean-up and MT shape&size re-work

### PR DESCRIPTION
**Pull request by ramjana**
- cleaned up localReadDo ;  
   * simplified f16/bf16 and f32 code
   * re-enabled mulitple offset(s) local read instruction for future use case
- LRO calculation
   * cleaned- up code
   * re-enabled multi-wave LRO  calculation
   * support of  miWG0 > wavefrontWidth 
   * local read optimization for matrixB=1 case (single fetch for both Kidx(s) 
 - computeStoreVgprs
   * update for miWG0 > wavefrontWidth 

- added xdlops test cases 
 - added test case for xdlops instructions

### CODE REVIEW

File: Tensile/KernelWriterAssembly.py [hunk 1]
```diff
@@ -5211,30 +5208,237 @@ def lraFinalOffset(self, kernel, tP):
...
       if kernel["VectorWidth"] > 1 and not kernel["MatrixInstruction"]:
         kStr += staticMultiply(vgpr(tP["gpr"]["lro"]), 
             vgpr(tP["gpr"]["lro"]), \
             kernel["VectorWidth"], sgpr(tmpSgpr))
+
+      #if kernel["MatrixInstruction"] and not kernel["ProblemType"]["DataType"].isHalf():
+      # miWG0 requiriing > 1 SIMD case
+      if kernel["MatrixInstruction"]:
+        if tc == "A":  
+            #sub-case 1: MatrixInstB != case
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:5240]: Should there be sub-case 2 for other sizes?

File: Tensile/KernelWriterAssembly.py [hunk 2]
```diff
@@ -5211,30 +5208,237 @@ def lraFinalOffset(self, kernel, tP):
...
+                hex(globalParameters["WavefrontWidth"]-1), \
+                vgpr(dividendReg), \
+                "WaveLaneoffset =  %s & %s-1"%(vgpr(dividendReg),hex(globalParameters["WavefrontWidth"])))
+            #optimization : fetch both MiK+0, MiK+B in single local Read
+            kStr += inst("v_lshrrev_b32", \
+                vgpr(rReg), \
+                log2(kernel["MatrixInstM"]), \
+                vgpr(rReg), \
+                "N_KIdx = v1 >> %s"%(log2(kernel["MatrixInstM"])))
+            if kernel["ProblemType"]["DataType"].isBFloat16() :
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:5413]: 16-bit code is duplicated in a few places, could be consolidated a bit

File: Tensile/KernelWriterAssembly.py [hunk 3]
```diff
@@ -7611,56 +7815,92 @@ def localReadDo(self, kernel, bufferIdx, iui, epsi, uIdx, tP):
...
+    # numReadsPerVector = take care of number of local reads required per instruction
+    # re-insert numberofOffset loop ; next optimization to fetch different colIdx/rowIdx for broadcast/swizzling option
+    # next clean-up 
+    #   transposeLDS case 
+
     numVectorsPerTile = (kernel["ThreadTile%u"%tP["tensorIdx"]]//kernel["VectorWidth"])
     if (kernel["MatrixInstruction"]):
-      numVectorsPerTile = 1 # TODO Fix for > tile
+      if tc == "A":
+        numVectorsPerTile = kernel["ThreadTile0"] # TODO Fix for > tile
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:7831]: Does TODO still apply? I think that note can be removed now

File: Tensile/KernelWriterAssembly.py [hunk 4]
```diff
@@ -7611,56 +7815,92 @@ def localReadDo(self, kernel, bufferIdx, iui, epsi, uIdx, tP):
...
+      else:
+        numVectorsPerTile = kernel["ThreadTile1"] // kernel["MatrixInstN"]
     #print "numVectorsPerTile", numVectorsPerTile
     numReadsPerVector = (kernel["VectorWidth"] * tP["bpe"]) // (blockWidth*4) # bytes/register
     if kernel["MatrixInstruction"]:
-      if tc == "A":
-        numReadsPerVector = kernel["ThreadTile0"] # TODO controls instruction tile shape, recalc if definition changes
-      else:
-        numReadsPerVector = kernel["ThreadTile1"] // kernel["MatrixInstN"]
+      #numReadsPerVector = 1
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:7837]: Can remove commented code here

File: Tensile/KernelWriterAssembly.py [hunk 5]
```diff
@@ -7611,56 +7815,92 @@ def localReadDo(self, kernel, bufferIdx, iui, epsi, uIdx, tP):
...
+              paramList.append(int(offset))
+            if tP["isB"]:
+              if kernel["MatrixInstB"] == 1 or  kernel["MIWG0"] > globalParameters["WavefrontWidth"] :
+                outputBlocks = 1
+              else:
+                outputBlocks = globalParameters["WavefrontWidth"] // (kernel["MIWG0"])
+                #TODO: colBlocks are  non-contiguous col block0 from simd0,simd1,simd2, simd col block1 from simd0,simd1,...
+                  # StoreoffsetCalculation requireFix
+               
+              offset = ((((rIdx * kernel["MacroTile%u" % tIdx] + kernel["LdsPad%s"%tc]) + (outputBlocks*(vIdx*numOffsets+oIdx)*kernel["MatrixInstN"]) \
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:7895]: Duplicate of above

File: Tensile/KernelWriterAssembly.py [hunk 6]
```diff
@@ -7611,56 +7815,92 @@ def localReadDo(self, kernel, bufferIdx, iui, epsi, uIdx, tP):
...
+            if rIdx%2 == 0:
               paramList.append(destVgpr)
             else:
               paramList.append(tmpVgpr)
             paramList.append(vgpr("LocalReadAddr%s"%tc))
-            offset = ((rIdx * kernel["MacroTile%u" % tIdx] // kernel["ThreadTile%u" % tIdx] + kIdx*(kernel["MacroTile%s"%tc]+kernel["LdsPad%s"%tc]) + tP["localReadOffset"]) \
-                 *tP["bpe"]+tP["localReadSwapByteOffset"])//offsetMultiplier
-            paramList.append(int(offset))
-            oIdx = 0
+            if tP["isA"]:
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:7877]: This section looks like it could be simplified

File: Tensile/KernelWriterAssembly.py [hunk 7]
```diff
@@ -7708,11 +7948,26 @@ def localReadDo(self, kernel, bufferIdx, iui, epsi, uIdx, tP):
             if kernel["MatrixInstruction"]:
               if not kernel["TransposeLDS"] or (kernel["ProblemType"]["TLU%s"%tP["tensorChar"]] and (kernel["TransposeLDS"] == 1)):
                 tIdx = tP["tensorIdx"]
-                readOffsetWidth = kernel["MacroTile%u" % tIdx] // kernel["ThreadTile%u" % tIdx]
+                #readOffsetWidth = kernel["MacroTile%u" % tIdx] // kernel["ThreadTile%u" % tIdx]
+                if tP["isA"]:
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:7952]: Looks like this is duplicated from above

File: Tensile/KernelWriterAssembly.py [hunk 8]
```diff
@@ -8663,6 +8922,24 @@ def computeStoreVgprs(self, kernel, divisor, tid0Scale, tid1Scale):
...
+        # 1 row-block per Instruction
+        if  kernel["MatrixInstB"] == 1:
+          kStr += inst("v_lshrrev_b32", vgpr(tmpV4), log2(globalParameters["WavefrontWidth"]),vgpr(tid0),"Row-BlockId in macroTile0")
+          #kStr += inst("v_lshlrev_b32", vgpr(tmpV4), log2(kernel["MatrixInstM"]*kernel["ThreadTile0"]),vgpr(tmpV4),"Row-BlockId in macroTile0")
+          kStr += inst("s_mov_b32", sgpr(tmpSgpr), hex(kernel["MatrixInstM"]*kernel["ThreadTile0"]), "")
+          kStr += inst("v_mul_lo_u32", vgpr(tmpV4), sgpr(tmpSgpr),vgpr(tmpV4),"Row-BlockId in macroTile0")
+          kStr += inst("v_and_b32", vgpr(tid0), hex(globalParameters["WavefrontWidth"]-1), vgpr(tid0), "WaveFrontLaneId")
+       
+        # 2 row-blocks per Instruction
+        if kernel["MIWG0"] > globalParameters["WavefrontWidth"] and kernel["MatrixInstB"] != 1:
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:8938]: elif could make this clearer to read

File: Tensile/KernelWriterAssembly.py [hunk 9]
```diff
@@ -8663,6 +8922,24 @@ def computeStoreVgprs(self, kernel, divisor, tid0Scale, tid1Scale):
...
+          kStr += inst("v_lshrrev_b32", vgpr(tmpV4), log2(globalParameters["WavefrontWidth"]),vgpr(tid0),"Row-BlockId in macroTile0")
+          kStr += inst("s_mov_b32", sgpr(tmpSgpr), hex(2*kernel["MatrixInstM"]*kernel["ThreadTile0"]),"")
+          kStr += inst("v_mul_lo_u32", vgpr(tmpV4), sgpr(tmpSgpr),vgpr(tmpV4),"Row-BlockId in macroTile0")
+          kStr += inst("v_and_b32", vgpr(tid0), hex(globalParameters["WavefrontWidth"]-1), vgpr(tid0), "WaveFrontLaneId")
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:8942]: This block looks the same as the one above, are there any differences? If there are minor differences, could probably still be combined for clarity

File: Tensile/KernelWriterAssembly.py [hunk 10]
```diff
@@ -7611,56 +7815,92 @@ def localReadDo(self, kernel, bufferIdx, iui, epsi, uIdx, tP):
...
     #totalReads = (kernel["ThreadTile%u"%tP["tensorIdx"]]/blockWidth) / numOffsets
     valuIdx = 0
+
+    # cleanup code 
+    # simplify both f32, half/float16 code ; no need for branch (next optimization)
+    # remove unnecessary numReadsAlongK (numReadPerVector should take care of that)
+    # numVectorsPerTile = take care of numer of instructions per wave
+    # numReadsPerVector = take care of number of local reads required per instruction
+    # re-insert numberofOffset loop ; next optimization to fetch different colIdx/rowIdx for broadcast/swizzling option
+    # next clean-up
```
Comment (rosenrodt) [inline Tensile/KernelWriterAssembly.py:7825]: Agree with removing the extra var `numReadsAlongK` and half-type specific branch, as the original concept `numVectorsPerTile/numReadsPerVector/numOffsets` cover this already

File: Tensile/KernelWriterAssembly.py [hunk 11]
```diff
@@ -7673,10 +7913,10 @@ def localReadDo(self, kernel, bufferIdx, iui, epsi, uIdx, tP):
...
             
-            isHighBits = 0 if kIdx % 2 == 0 else 1
+            isHighBits = 0 if rIdx % 2 == 0 else 1
             localReadCode.addCode(Code.LocalReadInst(instruction.toCodeInst(paramTuple, 0, isHighBits), comment))
 
-            if kIdx % 2 == 1:
+            if rIdx % 2 == 1:
               # pack 2 half-dword into one
               if kernel["MatrixInstruction"] and \
                  kernel["AssertSummationElementMultiple"] % kernel["MatrixInstK"] != 0 and \
```
Comment (rosenrodt) [inline Tensile/KernelWriterAssembly.py:7922]: arbitrary-k size support has been put in `mfmaIter()` in upstream develop so this if-conditional can be gone if you do a rebase

File: Tensile/KernelWriterAssembly.py (hunk 8 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py:8938]: changed to elif

File: Tensile/KernelWriterAssembly.py (hunk 9 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py:8942]: combined both into single block..

File: Tensile/KernelWriterAssembly.py [hunk 12]
```diff
@@ -5006,9 +5006,16 @@ def lraTileAssignmentA(self, kernel, tP):
...
+    ## we need to mulitply * instSplit to get divisor for InstB==1
+    ## each simd can only produce one matrixInstMxmatrixInstN
+    ## if subgroup0 > MatrixInstM if need multiple SIMD(S) to performe output space 
+    if kernel["MatrixInstruction"] and kernel["MatrixInstB"] == 1:
+       divisor = divisor * kernel["InstSplit"]
+    ## not sure why we need below code for f32 
+    ## pack =1 for f32
+    #if kernel["MatrixInstruction"] and not (kernel["ProblemType"]["DataType"].isHalf() or kernel["ProblemType"]["DataType"].isBFloat16()):
+    #  pack = 4 // tP["bpe"]
+    #  divisor //= pack
```
[the remark this answers is not in the document]
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: done

File: Tensile/KernelWriterAssembly.py [hunk 13]
```diff
@@ -5030,15 +5037,12 @@ def lraTileAssignmentB(self, kernel, tP):
         % (self.commentPrefix, tP["tileChar"], tP["tileChar"], \
         tP["tileChar"], self.commentSuffix, self.endLine)
     divisor = kernel["SubGroup1"]
-    if kernel["MatrixInstruction"]:
-      divisor = kernel["MatrixInstN"] if ((kernel["MacroTile1"] // 4) < kernel["MatrixInstN"]) else kernel["MacroTile1"] // 4  # ABlocks
-      divisor //= kernel["ThreadTile1"] // kernel["MatrixInstN"]
+    #if kernel["MatrixInstruction"]:
+    #  divisor = kernel["MatrixInstN"] if ((kernel["MacroTile1"] // 4) < kernel["MatrixInstN"]) else kernel["MacroTile1"] // 4  # ABlocks
+    #  divisor //= kernel["ThreadTile1"] // kernel["MatrixInstN"]
```
[the remark this answers is not in the document]
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: done

File: Tensile/KernelWriterAssembly.py (hunk 1 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: subcase is below numBlocks=1

File: Tensile/KernelWriterAssembly.py (hunk 2 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: done.

File: Tensile/KernelWriterAssembly.py (hunk 5 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py:8066]: localReadDo requires one more round simplification.. may be next version

File: Tensile/KernelWriterAssembly.py (hunk 7 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py:8115]: cleaned up some part of code more clean-up in next version

File: Tensile/KernelWriterAssembly.py (hunk 8 above)
Comment (ramjana) [reply Tensile/KernelWriterAssembly.py]: latest code brought back both version . there is subtle change in 2-row blocks per instruction

**Pull request closed without merging**